Implement basic stack-overview command

nicl opened the pull request:
This is a work in progress. A few things to consider/change:

* ~migrate the 'overview' command to use the new 'doForAll' pattern~ (DONE)
* support CSV output for the 'list-inactive' command
* tweak time periods

Commit:
Implement basic list-inactive command

This is a work in progress. A few things to consider/change:

* migrate the 'overview' command to use the new 'doForAll' pattern
* support CSV output for the 'list-inactive' command
* tweak time periods

nicl commented on src/index.ts:
```diff
@@ -33,5 +42,21 @@ void yargs(hideBin(process.argv))
+    async (argv) => {
+      const logLevel = getLogLevel(argv["log-level"] as string);
+      configureLogging(logLevel);
+      await doForAll(Config.AWS_PROFILES, Config.AWS_REGIONS, runListInactive);
```
(Suggesting we migrate overview to use this pattern too for handling the profiles/regions combos.)

nicl commented on src/list-inactive.ts:
```diff
@@ -0,0 +1,186 @@
+  NoRecentAsgDeployment: "Stack ASG has not been deployed to in the last month.",
+  NoRecentLambdaDeployment: "Stack Lambda has not been updated in the last 6 months.",
+  TestInName: "Stack name contains 'Test', which suggests it may be temporary.",
+};
```
Above are the key checks at the moment. The time periods are somewhat arbitrary, though the 1 month for ASGs is motivated by our 2 weeks for AMI updates target (with some breathing room).

nicl replied:
Above are the key checks at the moment. The time periods are somewhat arbitrary, though the 1 month for ASGs is motivated by our 2 weeks for AMI updates target (with some breathing room).

nicl commented on src/list-inactive.ts:
```diff
@@ -0,0 +1,186 @@
+
+type InactiveStatus = keyof typeof Inactive;
+
+const marshallTxt = (profile: string, region: string, stacks: Map<StackMetadata, InactiveStatus>): string => {
```
My thinking here is to support a `format=csv|text` flag for all commands. But I haven't implemented CSV here yet.

nicl commented on src/list-inactive.ts:
```diff
@@ -0,0 +1,186 @@
+  NoRecentUpdate: "Stack has not been updated at all in the last 2 years.",
+  NoRecentAsgDeployment: "Stack ASG has not been deployed to in the last month.",
+  NoRecentLambdaDeployment: "Stack Lambda has not been updated in the last 6 months.",
+  TestInName: "Stack name contains 'Test', which suggests it may be temporary.",
```
Above are the key checks at the moment. The time periods are somewhat arbitrary, though the 1 month for ASGs is motivated by our 2 weeks for AMI updates target (with some breathing room).

Commit:
Extend doForAll to stackReport

akash1810 commented on package.json:
```diff
@@ -25,6 +25,9 @@
   "dependencies": {
+    "@aws-sdk/client-auto-scaling": "^3.76.0",
+    "@aws-sdk/client-ec2": "^3.76.0",
+    "@aws-sdk/client-lambda": "^3.79.0",
```
Should these versions all match? (including `@aws-sdk/client-cloudformation` and `@aws-sdk/credential-provider-ini` in dev deps?)

akash1810 reviewed (commented):
Great start! I think we could benefit by thinking how this information will be used - if we're expecting teams to execute this script themselves or if we're to produce a spreadsheet with findings? If the latter, I think adding this information as an additional column to the other report would be simpler?

jacobwinch commented on src/list-inactive.ts:
```diff
@@ -0,0 +1,186 @@
+  const asgId = describeStackOutput.StackResourceDetail?.PhysicalResourceId;
+  const describeScalingActivitiesCmd = new DescribeScalingActivitiesCommand({ AutoScalingGroupName: asgId });
+  const describeScalingActivitiesOutput = await clients.asgClient.send(describeScalingActivitiesCmd);
+  const activities = describeScalingActivitiesOutput.Activities;
```
Based on [these docs](https://docs.aws.amazon.com/cli/latest/reference/autoscaling/describe-scaling-activities.html#:~:text=Describes%20scaling%20activity%2C%20which%20is,size%20or%20replacing%20an%20instance.), I wonder if this will also contain scaling events that are not related to a deployment (e.g. replacing an unhealthy instance or scaling on CPU usage)? I would guess that the latter is rare but the former may be reasonably likely for older services?

If so, perhaps we can filter based on the desired capacity being increased so that we don't accidentally include older services that are cycling boxes?

nicl replied:
It's a great point, at the moment this is not really doing the right thing. I think we can filter activities to those with a Cause including the text 'user request' as the following formats exist:

```
At 2022-05-03T11:15:36Z a user request explicitly set group desired capacity changing the desired capacity from 1 to 2. At 2022-05-03T11:15:49Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 1 to 2.

At 2022-05-04T11:07:38Z instance i-0dfa9f52d297e23b8 was taken out of service in response to a user request, shrinking the capacity from 2 to 1.
```

It's a bit brittle, but should work.

Commit:
Filter activities to user requests

Commit:
Bump and align aws deps

Commit:
Add 'backStackState' check to list-inactive

This is directly useful but also required as otherwise calls to
lookup resources (e.g. ASGs and Lambdas) may fail when the stack is
in a bad state.

Commit:
Refactor to separate 'stack' from 'resource' overviews

Stack -> for stack-related reporting
Resource -> for resource-level reporting

Note, there is inevitable overlap around core Stack metadata, which
is useful to include in both.

Console output has also been simplified to output just the full
CSV paths.

nicl commented:
@akash1810 have updated following chats to output CSV and split into `resource-overview` and `stack-overview` commands. (Note, it's a bit confusing because `resource-overview` is the previous `stack-report` command.)

Example output of the new `stack-overview` command for Frontend can be found here:

https://docs.google.com/spreadsheets/d/15y9mTOyldsIjv6kg_JsmYmxG60JL_uv-bX01H6-tDTE/edit?usp=sharing

It feels useful to flag stacks that haven't seen love in a long time and need some attention/review - there are definitely several examples of this in Frontend.

akash1810 commented on src/stack-overview.ts:
```diff
@@ -0,0 +1,254 @@
+  };
+
+  const stacks = await getStacks([profile], [region], preferCache);
+  const clients = {
+    cfnClient: new CloudFormationClient({
+      region,
+      maxAttempts: Config.SDK_MAX_ATTEMPTS,
+      credentials: fromIni({ profile }),
+    }),
+    asgClient: new AutoScalingClient({
+      region,
+      maxAttempts: Config.SDK_MAX_ATTEMPTS,
+      credentials: fromIni({ profile }),
+    }),
+    lambdaClient: new LambdaClient({
+      region,
+      maxAttempts: Config.SDK_MAX_ATTEMPTS,
+      credentials: fromIni({ profile }),
+    }),
+  };
```
Does this require more permissions than the CFN Read ones from before? I wonder if this repo is moving in the self-serve direction, rather than DevX produces a spreadsheet for others to consume?

nicl replied:
Hmm, it's a good point. I'm not sure how best to consume this to be honest. We could link it to the CDK getting started docs somehow?

akash1810 replied:
Happy to defer this.

akash1810 commented on src/stack-overview.ts:
```diff
@@ -0,0 +1,254 @@
+    badStackState,
+    noRecentUpdate,
+    testInName,
+    noRecentAsgDeployment.bind(null, clients),
+    noRecentLambdaDeployment.bind(null, clients),
```
Why `bind` here?

nicl replied:
I'm building a list of `StackMetadata => Promise<InactiveStatus>`, which can then be iterated, but some of the checks have additional args (the clients in this case) so bind is to partially apply them.

akash1810 commented:
> Example output of the new `stack-overview` command for Frontend can be found here:
> 
> https://docs.google.com/spreadsheets/d/15y9mTOyldsIjv6kg_JsmYmxG60JL_uv-bX01H6-tDTE/edit?usp=sharing

Looks great! We should filter out instances of stack sets in a future PR.
